diff -rupN minix/fs/mfs/read.c new_minix/fs/mfs/read.c
--- minix/fs/mfs/read.c	2017-05-31 19:17:52.616897426 +0200
+++ new_minix/fs/mfs/read.c	2017-05-31 23:57:05.059101728 +0200
@@ -269,6 +269,36 @@ int *completed;			/* number of bytes cop
 		}
 		return r;
 	} else {
+        
+        if (rw_flag == WRITING) {
+            // create buffer to get data from user space and be able to check if it's filled with zeros
+             char *buffer = calloc(sizeof(char), block_size);
+
+             if (buffer == NULL)
+                 panic("MFS, rw_chunk: allocating memory for checking if block is empty");
+
+             // copy bytes from user space to buffer
+             int r = sys_safecopyfrom(VFS_PROC_NR, gid, (vir_bytes) buf_off, 
+                                      (vir_bytes) (buffer), (size_t) chunk);
+
+             if (r != OK) {
+                free(buffer);
+                return r;
+             }
+
+             // check if buffer is filled with zeros
+             int is_empty = 1;
+             for (int i = 0; is_empty && i < block_size; i++) {
+                if (buffer[i] != 0)
+                    is_empty = 0;
+             }
+
+             free(buffer);
+
+             if (is_empty)
+                return r;
+        }
+ 
 		/* Writing to or peeking a nonexistent block.
 		 * Create and enter in inode.
 		 */
